[Documentation]: Polishing TimeSeries Best Practice Section (#115)
NeurodataWithoutBorders/nwbinspector: docs/best_practices/time_series.rst

Time Series
===========

When using :py:class:`~pynwb.base.TimeSeries` or any of its descendants, please ensure the following practices are followed.

Note that these extend to any general subtype, such as :py:class:`~pynwb.ecephys.ElectricalSeries`,

> :py:class:`~pynwb.behavior.SpatialSeries`, :py:class:`~pynwb.image.ImageSeries`, etc.

Reviewer: ```suggestion
When using :py:class:`~pynwb.base.TimeSeries` or any subtype (e.g. :py:class:`~pynwb.ecephys.ElectricalSeries`, :py:class:`~pynwb.behavior.SpatialSeries`, and :py:class:`~pynwb.image.ImageSeries`), please ensure the following practices are followed.
```

.. _best_practice_data_orientation:

Data Orientation
~~~~~~~~~~~~~~~~

The time dimension always goes first. In :py:attr:`~pynwb.base.TimeSeries.data`, the first dimension on the disk is always time.

Keep in mind that the dimensions are reversed in MatNWB, so in memory in MatNWB the time dimension must be last.

In PyNWB the order of the dimensions is the same in memory as on disk, so the time index should be first.

Check functions: :py:meth:`~nwbinspector.checks.time_series.check_data_orientation` and
:py:meth:`~nwbinspector.checks.time_series.check_timestamps_match_first_dimension`

.. _best_practice_unit_of_measurement:

Units of Measurement
~~~~~~~~~~~~~~~~~~~~

<<<<<<< HEAD
Time-related values should always in seconds. This include :py:attr:`~pynwb.base.TimeSeries.rate` (if applicable), which
should should be in Hz.
=======
Time-related values should always in seconds. This includes ``rate`` (if applicable), which should should be in Hz.

> >>>>>>> 09ce3dcf67d7be26d12a4710bc5ab252d9a1c791

Reviewer: unresolved conflict

Use chunking to optimize reading of large data for your use case.

By default, when using the HDF5 backend, :nwb-schema:ref:`sec-TimeSeries` ``data`` are stored on disk using
column-based ordering.

This means that if the ``data`` of a :nwb-schema:ref:`sec-TimeSeries` has multiple dimensions, then all data from a
single timestamp are stored contiguously on disk, followed by the next timestamp, and so on.

This storage scheme may be optimal for certain uses, such as slicing :nwb-schema:ref:`sec-TimeSeries` by time; however,
it may be sub-optimal for other uses, such as reading data from all timestamps for a particular value in the second or
third dimension.

This is especially important when writing NWBFiles that are intended to be uploaded to the
:dandi-archive:`DANDI Archive <>` for storage, sharing, and publication.

For more information about how to enable chunking and compression on your data, consult the
:pynwb-docs:`PyNWB tutorial <tutorials/advanced_io/h5dataio.html#chunking>` or the
`MatNWB instructions <https://neurodatawithoutborders.github.io/matnwb/tutorials/html/dataPipe.html#2>`_.

.. _best_practice_large_dataset_compression:

Compress Data
~~~~~~~~~~~~~

> Data writers can optimize the storage of large data arrays for particular uses by using compression applied to each
> chunk individually.
> 
> This is especially important when writing NWBFiles that are intended to be uploaded to the
> :dandi-archive:`DANDI Archive <>` for storage, sharing, and publication.
> 
> For more information about how to enable compression on your data, consult the

Reviewer: ```suggestion
Data writers can optimize the storage of large data arrays for particular uses by using compression applied to each
chunk individually. This is especially important when writing NWBFiles that are intended to be uploaded to the
:dandi-archive:`DANDI Archive <>` for storage, sharing, and publication. For more information about how to enable compression on your data, consult the
```

for other uses, such as reading data from all timestamps for a particular value in the second or third dimension.

This is especially important when writing NWBFiles that are intended to be uploaded to the 
:dandi-archive:`DANDI Archive <>` for storage, sharing, and publication.

For more information about how to enable chunking and compression on your data, consult the 
:pynwb-docs:`PyNWB tutorial <tutorials/advanced_io/h5dataio.html#chunking>` or the
`MatNWB instructions <https://neurodatawithoutborders.github.io/matnwb/tutorials/html/dataPipe.html#2>`_.

.. _best_practice_large_dataset_compression:

Compress Data
~~~~~~~~~~~~~

Data writers can optimize the storage of large data arrays for particular uses by using compression applied to each 
chunk individually.

This is especially important when writing NWBFiles that are intended to be uploaded to the 
:dandi-archive:`DANDI Archive <>` for storage, sharing, and publication.

For more information about how to enable compression on your data, consult the 
:pynwb-docs:`PyNWB tutorial <tutorials/advanced_io/h5dataio.html#compression-and-other-i-o-filters>` or the

> `MatNWB instructions <https://neurodatawithoutborders.github.io/matnwb/tutorials/html/dataPipe.html#2>`_

Author: This is the best I could find for the MatNWB side; is there something newer? I notice this is under the older `github.io`

In PyNWB the order of the dimensions is the same in memory as on disk, so the time index should be first.

Check functions: :py:meth:`~nwbinspector.checks.time_series.check_data_orientation` and
:py:meth:`~nwbinspector.checks.time_series.check_timestamps_match_first_dimension`

.. _best_practice_unit_of_measurement:

Units of Measurement
~~~~~~~~~~~~~~~~~~~~

> Time-related values should always in seconds. This include ``rate`` (if applicable), which should should be in Hz.

Reviewer: ```suggestion
Time-related values should always in seconds. This includes ``rate`` (if applicable), which should should be in Hz.
```

object with ``unit`` set to ``Volts``.

.. _best_practice_timestamps_ascending:

Breaks in Continuity
~~~~~~~~~~~~~~~~~~~~
TimeSeries data should generally be stored as one continuous stream as it was acquired, not by trial as is often
reshaped for analysis.

Data can be trial-aligned on-the-fly using the ``trials`` table.

Storing measured data as a continuous stream ensures that other users have access to the inter-trial data, and that we
can align the data within any specifiable window.

If you only have data spanning specific segments of time, then only include those timepoints in the data, see
:ref:`best_practice_regular_timestamps` for more information.

A primary implication is that the values in ``timestamps``, as well as the corresponding ordering of their indices
in the ``data`` array, should always be strictly increasing.

Check function: :py:meth:`~nwbinspector.checks.time_series.check_timestamps_ascending`

.. _best_practice_regular_timestamps:

Timestamps vs. Start & Rate
~~~~~~~~~~~~~~~~~~~~~~~~~~~

``TimeSeries`` allows you to specify time using either ``timestamps`` or ``rate`` together with ``starting_time``

> (which defaults to 0).
> 
> If the sampling rate is constant, then specify the ``rate`` and ``starting_time`` instead of writing the full ``timestamps`` vector.

Reviewer: ```suggestion
(which defaults to 0). If the sampling rate is constant, then specify the ``rate`` and ``starting_time`` instead of writing the full ``timestamps`` vector.
```
